pdg/src/info.rs

```diff
@@ -67,18 +67,483 @@ fn set_flow_info(g: &mut Graph) {
+        assert!(!info(&pdg, c2).unique);
+        assert!(!info(&pdg, c3).unique);
+    }
+
+    #[test]
+    fn unique_interleave_onesided() {
+        let mut g = Graph::default();
+        // let mut a = 0;   // A
+        // let b = &mut a;  // B1
+        // *b = 0;          // B2
+        // let c = &mut a;  // C1
+        // *c = 0;          // C2
+        // *b = 0;          // B3
+        //
+        // A
+        // +----.
+        // B1   |
+        // +-B2 |
+        // |    C1
+        // |    +-C2
+        // B3
+        let a = mk_addr_of_local(&mut g, 0_u32);
+        let b1 = mk_copy(&mut g, a);
+        let b2 = mk_store_addr(&mut g, b1);
+        let c1 = mk_copy(&mut g, a);
+        let c2 = mk_store_addr(&mut g, c1);
+        let b3 = mk_store_addr(&mut g, b1);
+
+        let pdg = build_pdg(g);
+        assert!(!info(&pdg, a).unique);
+        assert!(!info(&pdg, b1).unique);
+        assert!(!info(&pdg, b2).unique);
+        assert!(!info(&pdg, b3).unique);
+        assert!(!info(&pdg, c1).unique);
+        assert!(!info(&pdg, c2).unique);
+    }
+
+    #[test]
+    fn unique_sub_borrow() {
+        let mut g = Graph::default();
+        // let mut a = 0;   // A
+        // let b = &mut a;  // B1
+        // *b = 0;          // B2
+        // let c = &mut *b; // C1
+        // *c = 0;          // C2
+        // *b = 0;          // B3
+        //
+        // A
+        // |
+        // B1
+        // +-B2
+        // +----C1
+        // |    +-C2
+        // B3
+        let a = mk_addr_of_local(&mut g, 0_u32);
+        let b1 = mk_copy(&mut g, a);
+        let b2 = mk_store_addr(&mut g, b1);
+        let c1 = mk_copy(&mut g, b1);
+        let c2 = mk_store_addr(&mut g, c1);
+        let b3 = mk_store_addr(&mut g, b1);
+
+        let pdg = build_pdg(g);
+        assert!(info(&pdg, a).unique);
+        assert!(info(&pdg, b1).unique);
+        assert!(info(&pdg, b2).unique);
+        assert!(info(&pdg, b3).unique);
+        assert!(info(&pdg, c1).unique);
+        assert!(info(&pdg, c2).unique);
+    }
+
+    #[test]
+    fn unique_sub_borrow_bad() {
+        let mut g = Graph::default();
+        // let mut a = 0;   // A
+        // let b = &mut a;  // B1
+        // *b = 0;          // B2
+        // let c = &mut *b; // C1
+        // *c = 0;          // C2
+        // *b = 0;          // B3
+        // *c = 0;          // C3
+        //
+        // A
+        // |
+        // B1
+        // +-B2
+        // +----C1
+        // |    +-C2
+        // B3   |
+        //      C3
+        let a = mk_addr_of_local(&mut g, 0_u32);
+        let b1 = mk_copy(&mut g, a);
+        let b2 = mk_store_addr(&mut g, b1);
+        let c1 = mk_copy(&mut g, b1);
+        let c2 = mk_store_addr(&mut g, c1);
+        let b3 = mk_store_addr(&mut g, b1);
+        let c3 = mk_store_addr(&mut g, c1);
+
+        let pdg = build_pdg(g);
+        assert!(info(&pdg, a).unique);
+        assert!(!info(&pdg, b1).unique);
```

Comment:
i'm trying to make sense of these, why should the ancestors of `a` be non-unique if they're written to?

Reply:
after our conversation offline, the test case makes sense to me